bigtable/integration_test.go

```diff
@@ -2259,3 +2318,38 @@ func isDirectPathRemoteAddress(testEnv IntegrationEnv) (_ string, _ bool) {
 	// DirectPath ipv6 can use either ipv4 or ipv6 traffic.
 	return remoteIP, strings.HasPrefix(remoteIP, directPathIPV4Prefix) || strings.HasPrefix(remoteIP, directPathIPV6Prefix)
 }
+
+func blackholeOrAllowDirectPath(testEnv IntegrationEnv, t *testing.T, blackholeDP bool) {
+	blackholeDpv6Cmd := "sudo ip6tables -I INPUT -s 2001:4860:8040::/42 -j DROP && sleep 5 && echo blackholeDpv6"
+	blackholeDpv4Cmd := "sudo iptables -I INPUT -s 34.126.0.0/18 -j DROP && sleep 5 && echo blackholeDpv4"
+	allowDpv6Cmd := "sudo ip6tables -I INPUT -s 2001:4860:8040::/42 -j ACCEPT && sleep 5 && echo allowDpv6"
+	allowDpv4Cmd := "sudo iptables -I INPUT -s 34.126.0.0/18 -j ACCEPT && sleep 5 && echo allowDpv4"
```

Comment:
This makes me really nervous. I don't expect a test to need `sudo` or alter `iptables`. Is there another way to do this?

Reply:
The test itself is a bit tricky because we want to create a bad situation to make the client think it can not receive dp response (i.e., blackholed), so that we can test fallback. Previously in java, we modified the grpc netty library to achieve the blackhole (code here: https://github.com/googleapis/java-bigtable/blob/master/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java#L184), but this is very hacky and sorry I have no idea how to do a similar thing in Go.

Reply:
Agreed that this seems really problematic as-is.

I'm more familiar with the HTTP clients but if I were testing something similar there, I would look into either setting up a small mock server to hit for my test, or using a custom roundtripper at the transport layer to produce the behavior I was looking for. Is it possible to do something similar in grpc?

Reply:
In this test, we want to create a low level failure on TCP/IP, and grpc does not have any API to do that. Besides, a mock server is more like what we will use in unit tests, but this is an integration test and we want to make it as real as possible, so we may not want to use a mock server. 

When we are running the tests, we have the ownership of the VM that runs the test (create the VM, use `sudo` to setup the VM, run the test, destroy the VM..), so it will fine to use `sudo` without worrying about environment consistency.

Also notice that users are not supposed to run this test, since users are not supposed to know if the traffic is DirectPath or CFE. I added a check for `AttemptDirectPath` flag at the beginning of the test. Does this looks good?

Reply:
From @igorbernstein2, this go test might help us out: https://github.com/grpc/grpc-go/blob/ff1fc890e43ac77e922f53a2cef396b3c6a8f2a1/interop/grpclb_fallback/client.go#L78

Reply:
I agree with the feedback, tests should avoid mutating the machine state.
Ideally it should follow a pattern similar to the java impl:
https://github.com/googleapis/java-bigtable/blob/master/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java

Where the network failures are inject in process. It appears that grpc-go has some hooks to do something similar via the DialContext. the grpclb_fallback test that @kolea2 linked to, seems to use those hooks to test this exact scenario

Reply:
@kolea2 @igorbernstein2 Thanks for the comments! I read the grpclb_fallback test and still have some questions.

1. The Dialer is from Golang instead of grpc-go, see https://golang.org/pkg/net/#Dialer. I made a custom dialer to print the dial address, as in https://paste.googleplex.com/6601422297628672. But when I run the test, the test can pass, but no custom log was printed. Any idea why it is not working?

2. If we can make the custom dialer work, then how can I block the directpath ip address in the custom dialer? Any example or API that I can learn from?

3. 1. The grpclb_fallback test is in fact using shell command to blackhole ip, see https://github.com/grpc/grpc-go/blob/ff1fc890e43ac77e922f53a2cef396b3c6a8f2a1/interop/grpclb_fallback/client.go#L202-L207, which seems the same as what I want to do? And sorry I am still a bit confused why we should not use shell command in tests?

Thanks!